**Fin 585**  
**Diether**  
**Python/Pandas Introduction**<br><br>

**Instructions**

+ Please read through my notes, and run each of the code cells.

+ You can run a cell of code by pressing SHIFT and ENTER at the same time.


**1 Python/Pandas in Empirical Finance**

**1.1 Role of Python/Pandas in this Course**

+ Goal: Develop the Python/Pandas skills and tools necessary to engage in empirical research in Finance.

+ More specifically, use Python/Pandas to $\rightarrow$

  - Test economic models
  
  - Construct portfolios (container for financial assets)
  
  - Create and backtest trading strategies.
  
  - Estimate regressions: time series and panel regressions.
  
+ I will focus on the most important features and programming constructs in Python/Pandas to accomplish these goals.<br><br>


**1.2 Example: Portfolio Construction and Trading Strategies**

+ A core quant finance and academic skill is portfolio construction and backtesting.

+ All trading strategies are implemented as portfolios (container for financial assets).

+ Portfolio construction and backtesting can be broken into five general steps:

  1. Data preparation.

  2. Creation of the portfolio formation variable.

  3. Binning the stock return data based the formation variable.

  4. Portfolio creation.

  5. Estimating historical performance of the strategy.
  
+ Need to learn enough Python/Pandas so you can tackle each step for portfolio strategies you're interested in testing.<br><br>


**2 Why Python/Pandas?**

+ Why Python/Pandas? Why not R, SAS, Stata, or something else?

+ All of those languages are used in empirical finance research.

+ Python/Pandas has some important advantages:

  - Very popular in finance world now.

  - Well designed and popular general purpose programming language.
  
  - Free
  
  - Relatively easily to learn. 
  
  - Used in lots of different domains; it's not narrowly confined to the domain of quantitative finance or even scientific computing or data science.<br><br>
  

**3 Overview of Basic Concepts and Features**

+ Main purpose of this `notebook` is to introduce the `Pandas` `library`.

+ `Pandas` is the main library for this course.

+ Will overview core concepts and features of `pandas` for quant and academic finance.

+ Will cover the concepts and features with more detail as we move forward.<br><br>

**3.1 Accessing the Pandas Library**

+ To use the Pandas library we have to tell Python that we want access to it.

+ You make `pandas` accessible by using the `import` command.

+ When importing the pandas' library, you also associate the library with a namespace: **I use pd**<br>

  - Just convention<br>

  - Given the `pd` namespace $\rightarrow$ Pandas' functions looke like `pd.function`.
  
  - For example, `pd.read_csv`. 

  - `pd` namespace is not required, but is a strong convention.

  - Namespaces make it clear what library a certain function or command comes from if each library you use has it's own namespace.

+ **code for importing pandas:**


In [1]:
import pandas as pd

<br>

**3.2 Pandas core Data Structures: Dataframes and Series**

+ Core data structure/object $\rightarrow$ the **dataframe**.

+ Dataframe: container for holding rectangular array of mixed type data called a `dataframe`.

  - Columns: represent different variables (e.g, stock price or earnings of Google).

  - Rows: represent a given observation for those variables (e.g., January 2009 for Google).

+ Dataframe: programming equivalent of a spreadsheet.

+ Each column can be of a different type: integers, floating point numbers,  strings, or even imaginary numbers. <br><br>


**3.3 Dataframes: Store Data and Provide Useful Functions**

+ Pandas' provides programmers with many ways to create new data, transform and combine data, aggregate data, or display data. 

  - Example: Pandas' has a built in operator (`/`) that allows you to divide one column into the other column element by element.<br>

  - Example: Built in mean function that computes sample average of each column.
  
  - Higher level functions. For example, plotting functions
  
+ Many functions are built into the dataframe.

+ Built in functions called `methods` (name comes from object oriented programming language).<br><br>


**3.4 Summary of DataFrames**

+ Dataframe is an object the provides data storage and useful functions<br><br>


**3.5 Series**

+ `Series` in pandas' is the name for a single column of data.
  
+ If you grab one column of a dataframe, you're grabbing a series.

+ `Dataframes` and `Series` behave very similarly. For our purposes, it's mostly just be a technical distinction between a one dimensional and two dimensional array.<br><br>


**3.6 Importing Data and Creating a DataFrame:**

+ Getting data into a `Pandas'` dataframe is usually straight forward and easy. 

+ Pandas easily reads many different data formats: csv files, Excel files, SAS data files, Stata data files (dta), Feather data files, etc.

+ In this class, we mostly use csv files.

+ I will highlight other methods.<br><br>


**4. Example: Reading in A Little Financial Data**

+ Let's read in some data, and create a `dataframe` object.

+ Data to be read into a `dataframe` are annual balance sheet data for Amazon and Hormel.

+ Data are in a csv file $\rightarrow$ use the `read_csv` function.

+ The `read_csv` function will automatically create a `dataframe` object containing the data in the csv file.

+ The `read_csv` function has a lot of options and flexibility (take a look at the [Pandas' documentation for ead_csv](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html)), but often you don't use any of them (particularly for well formed or non-messy csv files). 

+ The `read_csv` function can, of course, read files stored on your local machine, but it also can read files stored remotely on a webserver; you just need to provide a URL. 

+ The code below calls pandas' `read_csv` function and then reads the csv located at the URL in quotes. After reading the file it creates a dataframe and assigns the dataframe to `df`.<br>

In [2]:
df = pd.read_csv('https://diether.org/prephd/01-intro.csv')

+ To read in data from non-csv formats you generally invoke a command very similar to `read_csv`. For example, you can read in a `Stata` datafile using the following:
```python
df = pd.read_stata('filename.dta')
```

+ Many other ways to create dataframes.

+ For example, you can convert core Python data structures (like `lists` or `dictionaries`) into `dataframes`.<br><br> 


**4.1 Displaying or Printing out the Data in a Dataframe**

+ The **Jupyter notebook** is a special environment where if you type the name of a dataframe (or other datatypes), it will display the default view of that object.

+ If the dataframe is small it will display all the data, and if it's large only a truncated view of the data will be displayed. 

+ If you just write a python program and run it outside of the jupyter notebook environment, then you need to use the `print` function to see any output.


In [3]:
df

tick    year   revenue      ebit       capx       debt        mktcap
0   AMZN  2000.0    2762.0   -663.57    134.760    2127.50  5.557991e+03
1   AMZN  2001.0    3122.4   -230.67     50.321    2156.10  4.038219e+03
2   AMZN  2002.0    3932.9    105.70     39.163    2245.70  7.327544e+03
3   AMZN  2003.0    5263.7    270.74     45.963    1925.40  2.122449e+04
4   AMZN  2004.0    6921.1    431.38     89.133    1847.40  1.814610e+04
5   AMZN  2005.0    8490.0    465.00    204.000    1521.00  1.961440e+04
6   AMZN  2006.0   10711.0    409.00    216.000    1267.00  1.633644e+04
7   AMZN  2007.0   14835.0    655.00    224.000    1344.00  3.853824e+04
8   AMZN  2008.0   19166.0    789.00    333.000     533.00  2.194784e+04
9   AMZN  2009.0   24509.0   1183.00    373.000     252.00  5.972688e+04
10  AMZN  2010.0   34204.0   1407.00    979.000     641.00  8.118000e+04
11  AMZN  2011.0   48077.0    867.00   1811.000    1415.00  7.876050e+04
12  AMZN  2012.0   61093.0    672.00   3785.000    3830.00  1.138950e+05
13  AMZN  2013.0   74452.0    691.00   3444.000    5181.00  1.830446e+05
14  AMZN  2014.0   88988.0    130.00   4893.000   12489.00  1.441216e+05
15  AMZN  2015.0  107010.0   2176.00   4589.000   14183.00  3.183442e+05
16  AMZN  2016.0  135990.0   4066.00   6737.000   15213.00  3.576880e+05
17  AMZN  2017.0  177870.0   3954.00  11955.000   37926.00  5.635348e+05
18  AMZN  2018.0  232890.0  12242.00  13427.000   39787.00  7.344168e+05
19  AMZN  2019.0  280520.0  14177.00  16861.000   63205.00  9.202243e+05
20  AMZN  2020.0  386060.0  22315.00  40140.000   87789.00  1.638236e+06
21  AMZN  2021.0  469820.0  24429.00  61053.000  122600.00  1.697179e+06
22  AMZN  2022.0  513980.0  12907.00  63645.000  146820.00  8.603280e+05
23  AMZN  2023.0  574780.0  36913.00  52729.000  142210.00  1.556169e+06
24   HRL  2000.0    3675.1    262.61    100.120     145.93  2.612287e+03
25   HRL  2001.0    4124.1    300.31     77.129     462.41  3.727245e+03
26   HRL  2002.0    3910.3    310.51     64.465     409.65  3.229432e+03
27   HRL  2003.0    4200.3    307.02     67.104     395.27  3.579150e+03
28   HRL  2004.0    4779.9    357.02     80.363     361.51  4.323667e+03
29   HRL  2005.0    5414.0    418.57    107.090     350.43  4.510167e+03
30   HRL  2006.0    5745.5    448.91    141.520     350.05  5.138992e+03
31   HRL  2007.0    6193.0    478.57    125.800     350.00  5.485607e+03
32   HRL  2008.0    6754.9    509.43    125.890     350.00  4.181068e+03
33   HRL  2009.0    6533.7    531.79     96.961     350.00  5.138035e+03
34   HRL  2010.0    7220.7    643.18     89.823       0.00  6.823936e+03
35   HRL  2011.0    7895.1    715.53     96.911     250.00  7.733644e+03
36   HRL  2012.0    8230.7    725.98    132.300     250.00  8.227331e+03
37   HRL  2013.0    8751.7    793.18    106.760     250.00  1.190993e+04
38   HRL  2014.0    9316.3    923.34    159.140     250.00  1.373554e+04
39   HRL  2015.0    9263.9   1084.50    144.060     250.00  2.089523e+04
40   HRL  2016.0    9523.2   1286.20    255.520     250.00  1.840760e+04
41   HRL  2017.0    9167.5   1241.10    221.290     250.00  1.927160e+04
42   HRL  2018.0    9545.7   1157.20    389.610     624.84  2.281656e+04
43   HRL  2019.0    9497.3   1140.60    293.840     250.00  2.412199e+04
44   HRL  2020.0    9608.5   1064.60    367.500    1088.10  2.516558e+04
45   HRL  2021.0   11386.0   1118.00    232.420    3371.90  2.648284e+04
46   HRL  2022.0   12459.0   1285.40    278.920    3346.10  2.488961e+04
47   HRL  2023.0   12110.0   1126.00    270.210    2468.00  1.755903e+04

<bf>

**Print function**

+ You can also explicitly print a dataframe out using python's print function.


In [4]:
print(df)

    tick    year   revenue      ebit       capx       debt        mktcap
0   AMZN  2000.0    2762.0   -663.57    134.760    2127.50  5.557991e+03
1   AMZN  2001.0    3122.4   -230.67     50.321    2156.10  4.038219e+03
2   AMZN  2002.0    3932.9    105.70     39.163    2245.70  7.327544e+03
3   AMZN  2003.0    5263.7    270.74     45.963    1925.40  2.122449e+04
4   AMZN  2004.0    6921.1    431.38     89.133    1847.40  1.814610e+04
5   AMZN  2005.0    8490.0    465.00    204.000    1521.00  1.961440e+04
6   AMZN  2006.0   10711.0    409.00    216.000    1267.00  1.633644e+04
7   AMZN  2007.0   14835.0    655.00    224.000    1344.00  3.853824e+04
8   AMZN  2008.0   19166.0    789.00    333.000     533.00  2.194784e+04
9   AMZN  2009.0   24509.0   1183.00    373.000     252.00  5.972688e+04
10  AMZN  2010.0   34204.0   1407.00    979.000     641.00  8.118000e+04
11  AMZN  2011.0   48077.0    867.00   1811.000    1415.00  7.876050e+04
12  AMZN  2012.0   61093.0    672.00   3785.000    

<br>

**4.2 Dataframes and Series**

+ Our `dataframe` is called `df`.

+ If we select a column from the `dataframe` it will be of type `Series`.

+ We select a column of a dataframe (a Series) by wrapping the column's name in quotes.


In [5]:
df['revenue']

0       2762.0
1       3122.4
2       3932.9
3       5263.7
4       6921.1
5       8490.0
6      10711.0
7      14835.0
8      19166.0
9      24509.0
10     34204.0
11     48077.0
12     61093.0
13     74452.0
14     88988.0
15    107010.0
16    135990.0
17    177870.0
18    232890.0
19    280520.0
20    386060.0
21    469820.0
22    513980.0
23    574780.0
24      3675.1
25      4124.1
26      3910.3
27      4200.3
28      4779.9
29      5414.0
30      5745.5
31      6193.0
32      6754.9
33      6533.7
34      7220.7
35      7895.1
36      8230.7
37      8751.7
38      9316.3
39      9263.9
40      9523.2
41      9167.5
42      9545.7
43      9497.3
44      9608.5
45     11386.0
46     12459.0
47     12110.0
Name: revenue, dtype: float64

In [6]:
df[['tick','year','revenue']]

tick    year   revenue
0   AMZN  2000.0    2762.0
1   AMZN  2001.0    3122.4
2   AMZN  2002.0    3932.9
3   AMZN  2003.0    5263.7
4   AMZN  2004.0    6921.1
5   AMZN  2005.0    8490.0
6   AMZN  2006.0   10711.0
7   AMZN  2007.0   14835.0
8   AMZN  2008.0   19166.0
9   AMZN  2009.0   24509.0
10  AMZN  2010.0   34204.0
11  AMZN  2011.0   48077.0
12  AMZN  2012.0   61093.0
13  AMZN  2013.0   74452.0
14  AMZN  2014.0   88988.0
15  AMZN  2015.0  107010.0
16  AMZN  2016.0  135990.0
17  AMZN  2017.0  177870.0
18  AMZN  2018.0  232890.0
19  AMZN  2019.0  280520.0
20  AMZN  2020.0  386060.0
21  AMZN  2021.0  469820.0
22  AMZN  2022.0  513980.0
23  AMZN  2023.0  574780.0
24   HRL  2000.0    3675.1
25   HRL  2001.0    4124.1
26   HRL  2002.0    3910.3
27   HRL  2003.0    4200.3
28   HRL  2004.0    4779.9
29   HRL  2005.0    5414.0
30   HRL  2006.0    5745.5
31   HRL  2007.0    6193.0
32   HRL  2008.0    6754.9
33   HRL  2009.0    6533.7
34   HRL  2010.0    7220.7
35   HRL  2011.0    7895.1
36   HRL  2012.0    8230.7
37   HRL  2013.0    8751.7
38   HRL  2014.0    9316.3
39   HRL  2015.0    9263.9
40   HRL  2016.0    9523.2
41   HRL  2017.0    9167.5
42   HRL  2018.0    9545.7
43   HRL  2019.0    9497.3
44   HRL  2020.0    9608.5
45   HRL  2021.0   11386.0
46   HRL  2022.0   12459.0
47   HRL  2023.0   12110.0

+ You typically must wrap the column's name in `' '` because most column names are stored as strings.

+ You will need to reference columns this way as long as the variable names aren't entirely numeric (e.g., an integer). 

+ In Python, can delimit strings by using either single (' ') or double quotes (" ").<br><br>

**4.3 Checking the Data Type**

+ In Python, there is a `type` function that returns the type of a variable or object.

In [7]:
type(df)

pandas.core.frame.DataFrame

In [8]:
type(df['revenue'])

pandas.core.series.Series

<br>

**4.4 Data creation** 

+ A new column in a dataframe is typically created using the assignment operator.

+ Like most programming languages, the assignment operator is just the equal sign (`=`) in Python.

+ For example, suppose I want to create a new column that measures profit margin. Profit margin is defined as the following (note, ebit is earnings before interest and taxes):
$$
\text{Profit Margin} = \frac{ebit}{revenue}
$$

+ Mathematical operations such as addition (+), subtraction (-), multiplication (*), or division (/) are all element by element operations between the dataframe columns that are addressed by the code.

+ Python/Pandas code for creating profit margin column in the dataframe.

In [9]:
df['profit_margin'] = df['ebit'] / df['revenue']
df

tick    year   revenue      ebit       capx       debt        mktcap  \
0   AMZN  2000.0    2762.0   -663.57    134.760    2127.50  5.557991e+03   
1   AMZN  2001.0    3122.4   -230.67     50.321    2156.10  4.038219e+03   
2   AMZN  2002.0    3932.9    105.70     39.163    2245.70  7.327544e+03   
3   AMZN  2003.0    5263.7    270.74     45.963    1925.40  2.122449e+04   
4   AMZN  2004.0    6921.1    431.38     89.133    1847.40  1.814610e+04   
5   AMZN  2005.0    8490.0    465.00    204.000    1521.00  1.961440e+04   
6   AMZN  2006.0   10711.0    409.00    216.000    1267.00  1.633644e+04   
7   AMZN  2007.0   14835.0    655.00    224.000    1344.00  3.853824e+04   
8   AMZN  2008.0   19166.0    789.00    333.000     533.00  2.194784e+04   
9   AMZN  2009.0   24509.0   1183.00    373.000     252.00  5.972688e+04   
10  AMZN  2010.0   34204.0   1407.00    979.000     641.00  8.118000e+04   
11  AMZN  2011.0   48077.0    867.00   1811.000    1415.00  7.876050e+04   
12  AMZN  2012.0   61093.0    672.00   3785.000    3830.00  1.138950e+05   
13  AMZN  2013.0   74452.0    691.00   3444.000    5181.00  1.830446e+05   
14  AMZN  2014.0   88988.0    130.00   4893.000   12489.00  1.441216e+05   
15  AMZN  2015.0  107010.0   2176.00   4589.000   14183.00  3.183442e+05   
16  AMZN  2016.0  135990.0   4066.00   6737.000   15213.00  3.576880e+05   
17  AMZN  2017.0  177870.0   3954.00  11955.000   37926.00  5.635348e+05   
18  AMZN  2018.0  232890.0  12242.00  13427.000   39787.00  7.344168e+05   
19  AMZN  2019.0  280520.0  14177.00  16861.000   63205.00  9.202243e+05   
20  AMZN  2020.0  386060.0  22315.00  40140.000   87789.00  1.638236e+06   
21  AMZN  2021.0  469820.0  24429.00  61053.000  122600.00  1.697179e+06   
22  AMZN  2022.0  513980.0  12907.00  63645.000  146820.00  8.603280e+05   
23  AMZN  2023.0  574780.0  36913.00  52729.000  142210.00  1.556169e+06   
24   HRL  2000.0    3675.1    262.61    100.120     145.93  2.612287e+03   
25   HRL  2001.0    4124.1    300.31     77.129     462.41  3.727245e+03   
26   HRL  2002.0    3910.3    310.51     64.465     409.65  3.229432e+03   
27   HRL  2003.0    4200.3    307.02     67.104     395.27  3.579150e+03   
28   HRL  2004.0    4779.9    357.02     80.363     361.51  4.323667e+03   
29   HRL  2005.0    5414.0    418.57    107.090     350.43  4.510167e+03   
30   HRL  2006.0    5745.5    448.91    141.520     350.05  5.138992e+03   
31   HRL  2007.0    6193.0    478.57    125.800     350.00  5.485607e+03   
32   HRL  2008.0    6754.9    509.43    125.890     350.00  4.181068e+03   
33   HRL  2009.0    6533.7    531.79     96.961     350.00  5.138035e+03   
34   HRL  2010.0    7220.7    643.18     89.823       0.00  6.823936e+03   
35   HRL  2011.0    7895.1    715.53     96.911     250.00  7.733644e+03   
36   HRL  2012.0    8230.7    725.98    132.300     250.00  8.227331e+03   
37   HRL  2013.0    8751.7    793.18    106.760     250.00  1.190993e+04   
38   HRL  2014.0    9316.3    923.34    159.140     250.00  1.373554e+04   
39   HRL  2015.0    9263.9   1084.50    144.060     250.00  2.089523e+04   
40   HRL  2016.0    9523.2   1286.20    255.520     250.00  1.840760e+04   
41   HRL  2017.0    9167.5   1241.10    221.290     250.00  1.927160e+04   
42   HRL  2018.0    9545.7   1157.20    389.610     624.84  2.281656e+04   
43   HRL  2019.0    9497.3   1140.60    293.840     250.00  2.412199e+04   
44   HRL  2020.0    9608.5   1064.60    367.500    1088.10  2.516558e+04   
45   HRL  2021.0   11386.0   1118.00    232.420    3371.90  2.648284e+04   
46   HRL  2022.0   12459.0   1285.40    278.920    3346.10  2.488961e+04   
47   HRL  2023.0   12110.0   1126.00    270.210    2468.00  1.755903e+04   

    profit_margin  
0       -0.240250  
1       -0.073876  
2        0.026876  
3        0.051435  
4        0.062328  
5        0.054770  
6        0.038185  
7        0.044152  
8        0.041167  
9        0.048268  
10       0.041136  
11       0.018034  
12       0.011000  

<br>

**4.5 If/then/else logic in Pandas**

+ `If/then/else` logic is important in all types of programming.

+ In `Python/Pandas`, you rarely write code that looks like classic `if/then/else` statements.

+ For example, many `Pandas` logical functions or statements are actually `if/then` statements with an implicit else.<br>

+ Data selection often involves if/then/else logic $\leftarrow$ in Pandas' jargon it's often called Boleen indexing.<br>

+ For example, we can use if/then/else logic to create a new variable that is `True` if the year is greater than 2010 and `False` otherwise. The logical statement looks like the following:

```
if (year is greater than 2010) then
   True
else
   False
```

+ Using python/pandas the code to implement the preceding logic is the following and it automatically creates a `Series` with `True` and `False` values based on the logical condtion that the year is greater than 2010:

In [10]:
df['year'] > 2010

0     False
1     False
2     False
3     False
4     False
5     False
6     False
7     False
8     False
9     False
10    False
11     True
12     True
13     True
14     True
15     True
16     True
17     True
18     True
19     True
20     True
21     True
22     True
23     True
24    False
25    False
26    False
27    False
28    False
29    False
30    False
31    False
32    False
33    False
34    False
35     True
36     True
37     True
38     True
39     True
40     True
41     True
42     True
43     True
44     True
45     True
46     True
47     True
Name: year, dtype: bool

+ We can also assign this new TRUE/FALSE variable to the dataframe: 

In [11]:
df['gt_2010'] = df['year'] > 2010
df

tick    year   revenue      ebit       capx       debt        mktcap  \
0   AMZN  2000.0    2762.0   -663.57    134.760    2127.50  5.557991e+03   
1   AMZN  2001.0    3122.4   -230.67     50.321    2156.10  4.038219e+03   
2   AMZN  2002.0    3932.9    105.70     39.163    2245.70  7.327544e+03   
3   AMZN  2003.0    5263.7    270.74     45.963    1925.40  2.122449e+04   
4   AMZN  2004.0    6921.1    431.38     89.133    1847.40  1.814610e+04   
5   AMZN  2005.0    8490.0    465.00    204.000    1521.00  1.961440e+04   
6   AMZN  2006.0   10711.0    409.00    216.000    1267.00  1.633644e+04   
7   AMZN  2007.0   14835.0    655.00    224.000    1344.00  3.853824e+04   
8   AMZN  2008.0   19166.0    789.00    333.000     533.00  2.194784e+04   
9   AMZN  2009.0   24509.0   1183.00    373.000     252.00  5.972688e+04   
10  AMZN  2010.0   34204.0   1407.00    979.000     641.00  8.118000e+04   
11  AMZN  2011.0   48077.0    867.00   1811.000    1415.00  7.876050e+04   
12  AMZN  2012.0   61093.0    672.00   3785.000    3830.00  1.138950e+05   
13  AMZN  2013.0   74452.0    691.00   3444.000    5181.00  1.830446e+05   
14  AMZN  2014.0   88988.0    130.00   4893.000   12489.00  1.441216e+05   
15  AMZN  2015.0  107010.0   2176.00   4589.000   14183.00  3.183442e+05   
16  AMZN  2016.0  135990.0   4066.00   6737.000   15213.00  3.576880e+05   
17  AMZN  2017.0  177870.0   3954.00  11955.000   37926.00  5.635348e+05   
18  AMZN  2018.0  232890.0  12242.00  13427.000   39787.00  7.344168e+05   
19  AMZN  2019.0  280520.0  14177.00  16861.000   63205.00  9.202243e+05   
20  AMZN  2020.0  386060.0  22315.00  40140.000   87789.00  1.638236e+06   
21  AMZN  2021.0  469820.0  24429.00  61053.000  122600.00  1.697179e+06   
22  AMZN  2022.0  513980.0  12907.00  63645.000  146820.00  8.603280e+05   
23  AMZN  2023.0  574780.0  36913.00  52729.000  142210.00  1.556169e+06   
24   HRL  2000.0    3675.1    262.61    100.120     145.93  2.612287e+03   
25   HRL  2001.0    4124.1    300.31     77.129     462.41  3.727245e+03   
26   HRL  2002.0    3910.3    310.51     64.465     409.65  3.229432e+03   
27   HRL  2003.0    4200.3    307.02     67.104     395.27  3.579150e+03   
28   HRL  2004.0    4779.9    357.02     80.363     361.51  4.323667e+03   
29   HRL  2005.0    5414.0    418.57    107.090     350.43  4.510167e+03   
30   HRL  2006.0    5745.5    448.91    141.520     350.05  5.138992e+03   
31   HRL  2007.0    6193.0    478.57    125.800     350.00  5.485607e+03   
32   HRL  2008.0    6754.9    509.43    125.890     350.00  4.181068e+03   
33   HRL  2009.0    6533.7    531.79     96.961     350.00  5.138035e+03   
34   HRL  2010.0    7220.7    643.18     89.823       0.00  6.823936e+03   
35   HRL  2011.0    7895.1    715.53     96.911     250.00  7.733644e+03   
36   HRL  2012.0    8230.7    725.98    132.300     250.00  8.227331e+03   
37   HRL  2013.0    8751.7    793.18    106.760     250.00  1.190993e+04   
38   HRL  2014.0    9316.3    923.34    159.140     250.00  1.373554e+04   
39   HRL  2015.0    9263.9   1084.50    144.060     250.00  2.089523e+04   
40   HRL  2016.0    9523.2   1286.20    255.520     250.00  1.840760e+04   
41   HRL  2017.0    9167.5   1241.10    221.290     250.00  1.927160e+04   
42   HRL  2018.0    9545.7   1157.20    389.610     624.84  2.281656e+04   
43   HRL  2019.0    9497.3   1140.60    293.840     250.00  2.412199e+04   
44   HRL  2020.0    9608.5   1064.60    367.500    1088.10  2.516558e+04   
45   HRL  2021.0   11386.0   1118.00    232.420    3371.90  2.648284e+04   
46   HRL  2022.0   12459.0   1285.40    278.920    3346.10  2.488961e+04   
47   HRL  2023.0   12110.0   1126.00    270.210    2468.00  1.755903e+04   

    profit_margin  gt_2010  
0       -0.240250    False  
1       -0.073876    False  
2        0.026876    False  
3        0.051435    False  
4        0.062328    False  
5        0.054770    False  
6        0.038185    False  
7        0.044152    False  
8        0.041167 

<br>

**4.6 Data selection** 

+ Based on if/then/else logic `Pandas` allows you to select only the rows or columns of a `dataframe` that you want.

+ Suppose you only want observations where the year is greater than 2010. Pandas allow us to index a dataframe's rows based on a logical condition or True/False Values.


In [12]:
df[df['gt_2010'] == True]

tick    year   revenue      ebit       capx       debt        mktcap  \
11  AMZN  2011.0   48077.0    867.00   1811.000    1415.00  7.876050e+04   
12  AMZN  2012.0   61093.0    672.00   3785.000    3830.00  1.138950e+05   
13  AMZN  2013.0   74452.0    691.00   3444.000    5181.00  1.830446e+05   
14  AMZN  2014.0   88988.0    130.00   4893.000   12489.00  1.441216e+05   
15  AMZN  2015.0  107010.0   2176.00   4589.000   14183.00  3.183442e+05   
16  AMZN  2016.0  135990.0   4066.00   6737.000   15213.00  3.576880e+05   
17  AMZN  2017.0  177870.0   3954.00  11955.000   37926.00  5.635348e+05   
18  AMZN  2018.0  232890.0  12242.00  13427.000   39787.00  7.344168e+05   
19  AMZN  2019.0  280520.0  14177.00  16861.000   63205.00  9.202243e+05   
20  AMZN  2020.0  386060.0  22315.00  40140.000   87789.00  1.638236e+06   
21  AMZN  2021.0  469820.0  24429.00  61053.000  122600.00  1.697179e+06   
22  AMZN  2022.0  513980.0  12907.00  63645.000  146820.00  8.603280e+05   
23  AMZN  2023.0  574780.0  36913.00  52729.000  142210.00  1.556169e+06   
35   HRL  2011.0    7895.1    715.53     96.911     250.00  7.733644e+03   
36   HRL  2012.0    8230.7    725.98    132.300     250.00  8.227331e+03   
37   HRL  2013.0    8751.7    793.18    106.760     250.00  1.190993e+04   
38   HRL  2014.0    9316.3    923.34    159.140     250.00  1.373554e+04   
39   HRL  2015.0    9263.9   1084.50    144.060     250.00  2.089523e+04   
40   HRL  2016.0    9523.2   1286.20    255.520     250.00  1.840760e+04   
41   HRL  2017.0    9167.5   1241.10    221.290     250.00  1.927160e+04   
42   HRL  2018.0    9545.7   1157.20    389.610     624.84  2.281656e+04   
43   HRL  2019.0    9497.3   1140.60    293.840     250.00  2.412199e+04   
44   HRL  2020.0    9608.5   1064.60    367.500    1088.10  2.516558e+04   
45   HRL  2021.0   11386.0   1118.00    232.420    3371.90  2.648284e+04   
46   HRL  2022.0   12459.0   1285.40    278.920    3346.10  2.488961e+04   
47   HRL  2023.0   12110.0   1126.00    270.210    2468.00  1.755903e+04   

    profit_margin  gt_2010  
11       0.018034     True  
12       0.011000     True  
13       0.009281     True  
14       0.001461     True  
15       0.020335     True  
16       0.029899     True  
17       0.022230     True  
18       0.052566     True  
19       0.050538     True  
20       0.057802     True  
21       0.051997     True  
22       0.025112     True  
23       0.064221     True  
35       0.090630     True  
36       0.088204     True  
37       0.090632     True  
38       0.099110     True  
39       0.117067     True  
40       0.135060     True  
41       0.135380     True  
42       0.121227     True  
43       0.120097     True  
44       0.110798     True  
45       0.098191     True  
46       0.103170     True  
47       0.092981     True

In [13]:
df[df['gt_2010']]

tick    year   revenue      ebit       capx       debt        mktcap  \
11  AMZN  2011.0   48077.0    867.00   1811.000    1415.00  7.876050e+04   
12  AMZN  2012.0   61093.0    672.00   3785.000    3830.00  1.138950e+05   
13  AMZN  2013.0   74452.0    691.00   3444.000    5181.00  1.830446e+05   
14  AMZN  2014.0   88988.0    130.00   4893.000   12489.00  1.441216e+05   
15  AMZN  2015.0  107010.0   2176.00   4589.000   14183.00  3.183442e+05   
16  AMZN  2016.0  135990.0   4066.00   6737.000   15213.00  3.576880e+05   
17  AMZN  2017.0  177870.0   3954.00  11955.000   37926.00  5.635348e+05   
18  AMZN  2018.0  232890.0  12242.00  13427.000   39787.00  7.344168e+05   
19  AMZN  2019.0  280520.0  14177.00  16861.000   63205.00  9.202243e+05   
20  AMZN  2020.0  386060.0  22315.00  40140.000   87789.00  1.638236e+06   
21  AMZN  2021.0  469820.0  24429.00  61053.000  122600.00  1.697179e+06   
22  AMZN  2022.0  513980.0  12907.00  63645.000  146820.00  8.603280e+05   
23  AMZN  2023.0  574780.0  36913.00  52729.000  142210.00  1.556169e+06   
35   HRL  2011.0    7895.1    715.53     96.911     250.00  7.733644e+03   
36   HRL  2012.0    8230.7    725.98    132.300     250.00  8.227331e+03   
37   HRL  2013.0    8751.7    793.18    106.760     250.00  1.190993e+04   
38   HRL  2014.0    9316.3    923.34    159.140     250.00  1.373554e+04   
39   HRL  2015.0    9263.9   1084.50    144.060     250.00  2.089523e+04   
40   HRL  2016.0    9523.2   1286.20    255.520     250.00  1.840760e+04   
41   HRL  2017.0    9167.5   1241.10    221.290     250.00  1.927160e+04   
42   HRL  2018.0    9545.7   1157.20    389.610     624.84  2.281656e+04   
43   HRL  2019.0    9497.3   1140.60    293.840     250.00  2.412199e+04   
44   HRL  2020.0    9608.5   1064.60    367.500    1088.10  2.516558e+04   
45   HRL  2021.0   11386.0   1118.00    232.420    3371.90  2.648284e+04   
46   HRL  2022.0   12459.0   1285.40    278.920    3346.10  2.488961e+04   
47   HRL  2023.0   12110.0   1126.00    270.210    2468.00  1.755903e+04   

    profit_margin  gt_2010  
11       0.018034     True  
12       0.011000     True  
13       0.009281     True  
14       0.001461     True  
15       0.020335     True  
16       0.029899     True  
17       0.022230     True  
18       0.052566     True  
19       0.050538     True  
20       0.057802     True  
21       0.051997     True  
22       0.025112     True  
23       0.064221     True  
35       0.090630     True  
36       0.088204     True  
37       0.090632     True  
38       0.099110     True  
39       0.117067     True  
40       0.135060     True  
41       0.135380     True  
42       0.121227     True  
43       0.120097     True  
44       0.110798     True  
45       0.098191     True  
46       0.103170     True  
47       0.092981     True

In [14]:
df[df['year'] > 2010]

tick    year   revenue      ebit       capx       debt        mktcap  \
11  AMZN  2011.0   48077.0    867.00   1811.000    1415.00  7.876050e+04   
12  AMZN  2012.0   61093.0    672.00   3785.000    3830.00  1.138950e+05   
13  AMZN  2013.0   74452.0    691.00   3444.000    5181.00  1.830446e+05   
14  AMZN  2014.0   88988.0    130.00   4893.000   12489.00  1.441216e+05   
15  AMZN  2015.0  107010.0   2176.00   4589.000   14183.00  3.183442e+05   
16  AMZN  2016.0  135990.0   4066.00   6737.000   15213.00  3.576880e+05   
17  AMZN  2017.0  177870.0   3954.00  11955.000   37926.00  5.635348e+05   
18  AMZN  2018.0  232890.0  12242.00  13427.000   39787.00  7.344168e+05   
19  AMZN  2019.0  280520.0  14177.00  16861.000   63205.00  9.202243e+05   
20  AMZN  2020.0  386060.0  22315.00  40140.000   87789.00  1.638236e+06   
21  AMZN  2021.0  469820.0  24429.00  61053.000  122600.00  1.697179e+06   
22  AMZN  2022.0  513980.0  12907.00  63645.000  146820.00  8.603280e+05   
23  AMZN  2023.0  574780.0  36913.00  52729.000  142210.00  1.556169e+06   
35   HRL  2011.0    7895.1    715.53     96.911     250.00  7.733644e+03   
36   HRL  2012.0    8230.7    725.98    132.300     250.00  8.227331e+03   
37   HRL  2013.0    8751.7    793.18    106.760     250.00  1.190993e+04   
38   HRL  2014.0    9316.3    923.34    159.140     250.00  1.373554e+04   
39   HRL  2015.0    9263.9   1084.50    144.060     250.00  2.089523e+04   
40   HRL  2016.0    9523.2   1286.20    255.520     250.00  1.840760e+04   
41   HRL  2017.0    9167.5   1241.10    221.290     250.00  1.927160e+04   
42   HRL  2018.0    9545.7   1157.20    389.610     624.84  2.281656e+04   
43   HRL  2019.0    9497.3   1140.60    293.840     250.00  2.412199e+04   
44   HRL  2020.0    9608.5   1064.60    367.500    1088.10  2.516558e+04   
45   HRL  2021.0   11386.0   1118.00    232.420    3371.90  2.648284e+04   
46   HRL  2022.0   12459.0   1285.40    278.920    3346.10  2.488961e+04   
47   HRL  2023.0   12110.0   1126.00    270.210    2468.00  1.755903e+04   

    profit_margin  gt_2010  
11       0.018034     True  
12       0.011000     True  
13       0.009281     True  
14       0.001461     True  
15       0.020335     True  
16       0.029899     True  
17       0.022230     True  
18       0.052566     True  
19       0.050538     True  
20       0.057802     True  
21       0.051997     True  
22       0.025112     True  
23       0.064221     True  
35       0.090630     True  
36       0.088204     True  
37       0.090632     True  
38       0.099110     True  
39       0.117067     True  
40       0.135060     True  
41       0.135380     True  
42       0.121227     True  
43       0.120097     True  
44       0.110798     True  
45       0.098191     True  
46       0.103170     True  
47       0.092981     True

<br>

**4.7 Creating a Sub Dataframe**

+ We can assign the smaller dataframe to a new dataframe with the following:

In [15]:
sub = df[df['year'] > 2010]
sub

tick    year   revenue      ebit       capx       debt        mktcap  \
11  AMZN  2011.0   48077.0    867.00   1811.000    1415.00  7.876050e+04   
12  AMZN  2012.0   61093.0    672.00   3785.000    3830.00  1.138950e+05   
13  AMZN  2013.0   74452.0    691.00   3444.000    5181.00  1.830446e+05   
14  AMZN  2014.0   88988.0    130.00   4893.000   12489.00  1.441216e+05   
15  AMZN  2015.0  107010.0   2176.00   4589.000   14183.00  3.183442e+05   
16  AMZN  2016.0  135990.0   4066.00   6737.000   15213.00  3.576880e+05   
17  AMZN  2017.0  177870.0   3954.00  11955.000   37926.00  5.635348e+05   
18  AMZN  2018.0  232890.0  12242.00  13427.000   39787.00  7.344168e+05   
19  AMZN  2019.0  280520.0  14177.00  16861.000   63205.00  9.202243e+05   
20  AMZN  2020.0  386060.0  22315.00  40140.000   87789.00  1.638236e+06   
21  AMZN  2021.0  469820.0  24429.00  61053.000  122600.00  1.697179e+06   
22  AMZN  2022.0  513980.0  12907.00  63645.000  146820.00  8.603280e+05   
23  AMZN  2023.0  574780.0  36913.00  52729.000  142210.00  1.556169e+06   
35   HRL  2011.0    7895.1    715.53     96.911     250.00  7.733644e+03   
36   HRL  2012.0    8230.7    725.98    132.300     250.00  8.227331e+03   
37   HRL  2013.0    8751.7    793.18    106.760     250.00  1.190993e+04   
38   HRL  2014.0    9316.3    923.34    159.140     250.00  1.373554e+04   
39   HRL  2015.0    9263.9   1084.50    144.060     250.00  2.089523e+04   
40   HRL  2016.0    9523.2   1286.20    255.520     250.00  1.840760e+04   
41   HRL  2017.0    9167.5   1241.10    221.290     250.00  1.927160e+04   
42   HRL  2018.0    9545.7   1157.20    389.610     624.84  2.281656e+04   
43   HRL  2019.0    9497.3   1140.60    293.840     250.00  2.412199e+04   
44   HRL  2020.0    9608.5   1064.60    367.500    1088.10  2.516558e+04   
45   HRL  2021.0   11386.0   1118.00    232.420    3371.90  2.648284e+04   
46   HRL  2022.0   12459.0   1285.40    278.920    3346.10  2.488961e+04   
47   HRL  2023.0   12110.0   1126.00    270.210    2468.00  1.755903e+04   

    profit_margin  gt_2010  
11       0.018034     True  
12       0.011000     True  
13       0.009281     True  
14       0.001461     True  
15       0.020335     True  
16       0.029899     True  
17       0.022230     True  
18       0.052566     True  
19       0.050538     True  
20       0.057802     True  
21       0.051997     True  
22       0.025112     True  
23       0.064221     True  
35       0.090630     True  
36       0.088204     True  
37       0.090632     True  
38       0.099110     True  
39       0.117067     True  
40       0.135060     True  
41       0.135380     True  
42       0.121227     True  
43       0.120097     True  
44       0.110798     True  
45       0.098191     True  
46       0.103170     True  
47       0.092981     True

<br>

**4.8 Deleting a Variable/Column of Data**

+ Very common to delete or remove columns.

+ Typically rely on the `drop` function.

+ For example, suppose I want to drop the `capx` column from the dataframe.


In [16]:
df.drop('capx',axis='columns')

tick    year   revenue      ebit       debt        mktcap  profit_margin  \
0   AMZN  2000.0    2762.0   -663.57    2127.50  5.557991e+03      -0.240250   
1   AMZN  2001.0    3122.4   -230.67    2156.10  4.038219e+03      -0.073876   
2   AMZN  2002.0    3932.9    105.70    2245.70  7.327544e+03       0.026876   
3   AMZN  2003.0    5263.7    270.74    1925.40  2.122449e+04       0.051435   
4   AMZN  2004.0    6921.1    431.38    1847.40  1.814610e+04       0.062328   
5   AMZN  2005.0    8490.0    465.00    1521.00  1.961440e+04       0.054770   
6   AMZN  2006.0   10711.0    409.00    1267.00  1.633644e+04       0.038185   
7   AMZN  2007.0   14835.0    655.00    1344.00  3.853824e+04       0.044152   
8   AMZN  2008.0   19166.0    789.00     533.00  2.194784e+04       0.041167   
9   AMZN  2009.0   24509.0   1183.00     252.00  5.972688e+04       0.048268   
10  AMZN  2010.0   34204.0   1407.00     641.00  8.118000e+04       0.041136   
11  AMZN  2011.0   48077.0    867.00    1415.00  7.876050e+04       0.018034   
12  AMZN  2012.0   61093.0    672.00    3830.00  1.138950e+05       0.011000   
13  AMZN  2013.0   74452.0    691.00    5181.00  1.830446e+05       0.009281   
14  AMZN  2014.0   88988.0    130.00   12489.00  1.441216e+05       0.001461   
15  AMZN  2015.0  107010.0   2176.00   14183.00  3.183442e+05       0.020335   
16  AMZN  2016.0  135990.0   4066.00   15213.00  3.576880e+05       0.029899   
17  AMZN  2017.0  177870.0   3954.00   37926.00  5.635348e+05       0.022230   
18  AMZN  2018.0  232890.0  12242.00   39787.00  7.344168e+05       0.052566   
19  AMZN  2019.0  280520.0  14177.00   63205.00  9.202243e+05       0.050538   
20  AMZN  2020.0  386060.0  22315.00   87789.00  1.638236e+06       0.057802   
21  AMZN  2021.0  469820.0  24429.00  122600.00  1.697179e+06       0.051997   
22  AMZN  2022.0  513980.0  12907.00  146820.00  8.603280e+05       0.025112   
23  AMZN  2023.0  574780.0  36913.00  142210.00  1.556169e+06       0.064221   
24   HRL  2000.0    3675.1    262.61     145.93  2.612287e+03       0.071457   
25   HRL  2001.0    4124.1    300.31     462.41  3.727245e+03       0.072818   
26   HRL  2002.0    3910.3    310.51     409.65  3.229432e+03       0.079408   
27   HRL  2003.0    4200.3    307.02     395.27  3.579150e+03       0.073095   
28   HRL  2004.0    4779.9    357.02     361.51  4.323667e+03       0.074692   
29   HRL  2005.0    5414.0    418.57     350.43  4.510167e+03       0.077313   
30   HRL  2006.0    5745.5    448.91     350.05  5.138992e+03       0.078132   
31   HRL  2007.0    6193.0    478.57     350.00  5.485607e+03       0.077276   
32   HRL  2008.0    6754.9    509.43     350.00  4.181068e+03       0.075416   
33   HRL  2009.0    6533.7    531.79     350.00  5.138035e+03       0.081392   
34   HRL  2010.0    7220.7    643.18       0.00  6.823936e+03       0.089074   
35   HRL  2011.0    7895.1    715.53     250.00  7.733644e+03       0.090630   
36   HRL  2012.0    8230.7    725.98     250.00  8.227331e+03       0.088204   
37   HRL  2013.0    8751.7    793.18     250.00  1.190993e+04       0.090632   
38   HRL  2014.0    9316.3    923.34     250.00  1.373554e+04       0.099110   
39   HRL  2015.0    9263.9   1084.50     250.00  2.089523e+04       0.117067   
40   HRL  2016.0    9523.2   1286.20     250.00  1.840760e+04       0.135060   
41   HRL  2017.0    9167.5   1241.10     250.00  1.927160e+04       0.135380   
42   HRL  2018.0    9545.7   1157.20     624.84  2.281656e+04       0.121227   
43   HRL  2019.0    9497.3   1140.60     250.00  2.412199e+04       0.120097   
44   HRL  2020.0    9608.5   1064.60    1088.10  2.516558e+04       0.110798   
45   HRL  2021.0   11386.0   1118.00    3371.90  2.648284e+04       0.098191   
46   HRL  2022.0   12459.0   1285.40    3346.10  2.488961e+04       0.103170   
47   HRL  2023.0   12110.0   1126.00    2468.00  1.755903e+04       0.092981   

    gt_2010  
0     False  
1     False  
2     False  
3     False  
4     False  

+ The preceding command, created a new `dataframe` with the `capx` column removed. 

+ Most `pandas` functions create a new dataframe.

+ To modify the original `dataframe` (df) we have to assign the `dataframe` created by the drop command back to `df`.

In [17]:
df = df.drop('capx',axis='columns')
df

tick    year   revenue      ebit       debt        mktcap  profit_margin  \
0   AMZN  2000.0    2762.0   -663.57    2127.50  5.557991e+03      -0.240250   
1   AMZN  2001.0    3122.4   -230.67    2156.10  4.038219e+03      -0.073876   
2   AMZN  2002.0    3932.9    105.70    2245.70  7.327544e+03       0.026876   
3   AMZN  2003.0    5263.7    270.74    1925.40  2.122449e+04       0.051435   
4   AMZN  2004.0    6921.1    431.38    1847.40  1.814610e+04       0.062328   
5   AMZN  2005.0    8490.0    465.00    1521.00  1.961440e+04       0.054770   
6   AMZN  2006.0   10711.0    409.00    1267.00  1.633644e+04       0.038185   
7   AMZN  2007.0   14835.0    655.00    1344.00  3.853824e+04       0.044152   
8   AMZN  2008.0   19166.0    789.00     533.00  2.194784e+04       0.041167   
9   AMZN  2009.0   24509.0   1183.00     252.00  5.972688e+04       0.048268   
10  AMZN  2010.0   34204.0   1407.00     641.00  8.118000e+04       0.041136   
11  AMZN  2011.0   48077.0    867.00    1415.00  7.876050e+04       0.018034   
12  AMZN  2012.0   61093.0    672.00    3830.00  1.138950e+05       0.011000   
13  AMZN  2013.0   74452.0    691.00    5181.00  1.830446e+05       0.009281   
14  AMZN  2014.0   88988.0    130.00   12489.00  1.441216e+05       0.001461   
15  AMZN  2015.0  107010.0   2176.00   14183.00  3.183442e+05       0.020335   
16  AMZN  2016.0  135990.0   4066.00   15213.00  3.576880e+05       0.029899   
17  AMZN  2017.0  177870.0   3954.00   37926.00  5.635348e+05       0.022230   
18  AMZN  2018.0  232890.0  12242.00   39787.00  7.344168e+05       0.052566   
19  AMZN  2019.0  280520.0  14177.00   63205.00  9.202243e+05       0.050538   
20  AMZN  2020.0  386060.0  22315.00   87789.00  1.638236e+06       0.057802   
21  AMZN  2021.0  469820.0  24429.00  122600.00  1.697179e+06       0.051997   
22  AMZN  2022.0  513980.0  12907.00  146820.00  8.603280e+05       0.025112   
23  AMZN  2023.0  574780.0  36913.00  142210.00  1.556169e+06       0.064221   
24   HRL  2000.0    3675.1    262.61     145.93  2.612287e+03       0.071457   
25   HRL  2001.0    4124.1    300.31     462.41  3.727245e+03       0.072818   
26   HRL  2002.0    3910.3    310.51     409.65  3.229432e+03       0.079408   
27   HRL  2003.0    4200.3    307.02     395.27  3.579150e+03       0.073095   
28   HRL  2004.0    4779.9    357.02     361.51  4.323667e+03       0.074692   
29   HRL  2005.0    5414.0    418.57     350.43  4.510167e+03       0.077313   
30   HRL  2006.0    5745.5    448.91     350.05  5.138992e+03       0.078132   
31   HRL  2007.0    6193.0    478.57     350.00  5.485607e+03       0.077276   
32   HRL  2008.0    6754.9    509.43     350.00  4.181068e+03       0.075416   
33   HRL  2009.0    6533.7    531.79     350.00  5.138035e+03       0.081392   
34   HRL  2010.0    7220.7    643.18       0.00  6.823936e+03       0.089074   
35   HRL  2011.0    7895.1    715.53     250.00  7.733644e+03       0.090630   
36   HRL  2012.0    8230.7    725.98     250.00  8.227331e+03       0.088204   
37   HRL  2013.0    8751.7    793.18     250.00  1.190993e+04       0.090632   
38   HRL  2014.0    9316.3    923.34     250.00  1.373554e+04       0.099110   
39   HRL  2015.0    9263.9   1084.50     250.00  2.089523e+04       0.117067   
40   HRL  2016.0    9523.2   1286.20     250.00  1.840760e+04       0.135060   
41   HRL  2017.0    9167.5   1241.10     250.00  1.927160e+04       0.135380   
42   HRL  2018.0    9545.7   1157.20     624.84  2.281656e+04       0.121227   
43   HRL  2019.0    9497.3   1140.60     250.00  2.412199e+04       0.120097   
44   HRL  2020.0    9608.5   1064.60    1088.10  2.516558e+04       0.110798   
45   HRL  2021.0   11386.0   1118.00    3371.90  2.648284e+04       0.098191   
46   HRL  2022.0   12459.0   1285.40    3346.10  2.488961e+04       0.103170   
47   HRL  2023.0   12110.0   1126.00    2468.00  1.755903e+04       0.092981   

    gt_2010  
0     False  
1     False  
2     False  
3     False  
4     False  

<br>

**5 More Advanced Concepts and Features for Later**

**5.1 The groupby/apply construct:**. 

+ The most important **programming idiom** or construct for this class is the `groupby/apply` construct.

+ Allows us to loop through the data and group observations in a `dataframe` together, and then apply a function or data transformation to each group.

+ For example, we often use it to group observations by date or to group all the observation of the same stock together. We then typically apply a function to the data that aggregates it or transforms it within these groups.

+ The `groupby/apply` construct allows us to accomplish the following with just one (or a few lines) of code:

  1. Logically **group** observations together based on some attribute of the data: for example, we could group stock data based on whether the company was big or small.

  2. **Apply** a function to the different groups. For example, we could compute the average number of analysts covering big versus small stocks.

+ The groupby/apply does a whole bunch of work for us behind the scene. It loops all the observations, categorizes the observations into the groups, and then applies the functions seperately to each group.<br><br>


**5.2 User-written functions:**

+ You will write your own custom (i.e., user written) functions to extend the functionality of the `groupby/apply` construct.<br>

+ For example, writing a custom function is sometimes and important part of implementing a portfolio formation criteria for a trading strategy.<br><br>


**5.3 Merging data** 

+ Merging data is a core part of the data preperation step from most empirical work or back testing of strategies.

+ You will learn how to merge dataframes together based on a single key or multiple keys.
